File: companion/bolumler/b02/ornek-01/cozum.R
veri <- read.csv("veri.csv", stringsAsFactors = FALSE,
                 fileEncoding = "UTF-8",
                 colClasses = c("character", "numeric", "numeric"))
stopifnot(identical(names(veri), c("okul_turu", "sinif", "puan")),
          nrow(veri) > 0, !anyNA(veri),
          all(veri$okul_turu %in% c("Devlet", "Ozel")),
          all(is.finite(veri$sinif)), all(is.finite(veri$puan)),
          all(veri$sinif %in% c(1, 2, 3)))
veri$okul_turu <- factor(veri$okul_turu, levels = c("Devlet", "Ozel"))
veri$sinif <- ordered(veri$sinif, levels = c(1, 2, 3))
sonuc <- data.frame(
  degisken = "veri", olcu = c("satir_sayisi", "degisken_sayisi"),
  deger = c(nrow(veri), ncol(veri))
)
for (alan in c("okul_turu", "sinif", "puan")) {
  sonuc <- rbind(sonuc, data.frame(
    degisken = alan, olcu = c("gecerli", "eksik"),
    deger = c(sum(!is.na(veri[[alan]])), sum(is.na(veri[[alan]])))
  ))
}
for (alan in c("okul_turu", "sinif")) {
  for (kategori in levels(veri[[alan]])) {
    frekans <- sum(veri[[alan]] == kategori)
    sonuc <- rbind(sonuc, data.frame(
      degisken = alan, olcu = paste0(kategori, c("_frekans", "_oran")),
      deger = c(frekans, frekans / nrow(veri))
    ))
  }
}
sonuc <- rbind(sonuc, data.frame(
  degisken = "puan", olcu = c("ortalama", "en_kucuk", "en_buyuk"),
  deger = c(mean(veri$puan), min(veri$puan), max(veri$puan))
))
for (okul in c("Devlet", "Ozel")) {
  sonuc <- rbind(sonuc, data.frame(
    degisken = "puan", olcu = paste0(okul, "_ortalama"),
    deger = mean(veri$puan[veri$okul_turu == okul])
  ))
}
cat(R.version.string, "\n")
print(sonuc, row.names = FALSE)
if ("--check" %in% commandArgs(trailingOnly = TRUE)) {
  beklenen <- read.csv("beklenen-sonuclar.csv", stringsAsFactors = FALSE)
  stopifnot(nrow(sonuc) == nrow(beklenen),
            all(sonuc$degisken == beklenen$degisken),
            all(sonuc$olcu == beklenen$olcu),
            isTRUE(all.equal(sonuc$deger, beklenen$deger,
                             tolerance = 1e-9, check.attributes = FALSE)))
  cat("DOGRULANDI:", nrow(beklenen), "kontrol degeri eslesiyor.\n")
}

File: companion/bolumler/b02/ornek-01/cozum.py
import argparse
import math
import platform

import pandas as pd


def hazirla(veri):
    if list(veri.columns) != ["okul_turu", "sinif", "puan"]:
        raise ValueError("CSV sutunlari ve sirasi veri sozluguyle uyusmuyor.")
    if veri.empty or veri.isna().any().any():
        raise ValueError("Bu pilot bos veri veya eksik hucre kabul etmiyor.")
    if not veri["okul_turu"].isin(["Devlet", "Ozel"]).all():
        raise ValueError("Okul turu Devlet veya Ozel olmali.")
    veri = veri.copy()
    for alan in ["sinif", "puan"]:
        veri[alan] = pd.to_numeric(veri[alan], errors="raise")
        if not veri[alan].map(math.isfinite).all():
            raise ValueError("Sayisal girdiler sonlu olmali.")
    if not veri["sinif"].isin([1, 2, 3]).all():
        raise ValueError("Sinif kodlari 1, 2 veya 3 olmali.")
    veri["okul_turu"] = pd.Categorical(
        veri["okul_turu"], categories=["Devlet", "Ozel"], ordered=False
    )
    veri["sinif"] = pd.Categorical(
        veri["sinif"], categories=[1, 2, 3], ordered=True
    )
    return veri


def hesapla(veri):
    veri = hazirla(veri)
    satirlar = [
        ("veri", "satir_sayisi", len(veri)),
        ("veri", "degisken_sayisi", len(veri.columns)),
    ]
    for alan in ["okul_turu", "sinif", "puan"]:
        satirlar.extend([
            (alan, "gecerli", veri[alan].count()),
            (alan, "eksik", veri[alan].isna().sum()),
        ])
    for alan in ["okul_turu", "sinif"]:
        for kategori in veri[alan].cat.categories:
            frekans = (veri[alan] == kategori).sum()
            satirlar.extend([
                (alan, str(kategori) + "_frekans", frekans),
                (alan, str(kategori) + "_oran", frekans / len(veri)),
            ])
    satirlar.extend([
        ("puan", "ortalama", veri["puan"].mean()),
        ("puan", "en_kucuk", veri["puan"].min()),
        ("puan", "en_buyuk", veri["puan"].max()),
    ])
    for okul in ["Devlet", "Ozel"]:
        grup = veri.loc[veri["okul_turu"] == okul, "puan"]
        satirlar.append(("puan", okul + "_ortalama", grup.mean()))
    return pd.DataFrame(satirlar, columns=["degisken", "olcu", "deger"])


def main():
    parser = argparse.ArgumentParser(description="B02 okul, sinif ve puan cozumu")
    parser.add_argument("--check", action="store_true",
                        help="Ozgun veri icin beklenen sonuclari denetle")
    args = parser.parse_args()
    veri = pd.read_csv("veri.csv", dtype={"okul_turu": "string"})
    sonuc = hesapla(veri)
    print(f"Python {platform.python_version()} | pandas {pd.__version__}")
    print(sonuc.to_string(index=False))
    if args.check:
        beklenen = pd.read_csv("beklenen-sonuclar.csv")
        anahtarlar = ["degisken", "olcu"]
        if not sonuc[anahtarlar].equals(beklenen[anahtarlar]):
            raise ValueError("Beklenen sonuc satirlari veya sirasi farkli.")
        for hesaplanan, hedef in zip(sonuc["deger"], beklenen["deger"]):
            if not math.isclose(hesaplanan, hedef, rel_tol=1e-9, abs_tol=1e-9):
                raise ValueError(f"Sonuc uyusmazligi: {hesaplanan} != {hedef}")
        print(f"DOGRULANDI: {len(beklenen)} kontrol degeri eslesiyor.")


if __name__ == "__main__":
    main()